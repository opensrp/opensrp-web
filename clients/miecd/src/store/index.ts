import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
    router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;

/** The initial store for the reveal web app */
const store = getConnectedStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
    store.replaceReducer(combine(reducers));
});

export default store;
