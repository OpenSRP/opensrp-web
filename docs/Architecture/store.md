# overview

This documentation will take you through the process of creating and working with:

    - reducers
    - actions
    - selectors

## Where do I find the store

the store reducer modules are found within src/store. The initial directory structure looks something like this: (Note: this structure only depicts the initial directory structure and will as expected contain more directories and files as more are added with time)

```
...
├── src
│   ├── ...
│   ├── store
│   │   ├── ducks
│   │   │    ├── superset.ts
│   │   │    ├── tests/
│   │   ├── tests
│   │   ├── index.ts
│   │   └── selectors.ts
...
```

A single reducer module will contain the afore mentioned parts i.e. its respective actions, action creators if any, its selectors and the reducer function

## Actions

An action can be any object with a type property, they are the only way provisioned for changing the state in a redux store.

### Adding Actions

Its a common convention as well as recommended one to Use type constants, [read more on this here]().

steps:

- Add An interface describing the action structure: redux exposes an `AnyAction` type that can be extended to write custom action interfaces
- Create a general actiontypes to be used to type check all actions that will be passed to the reducer function as an argument. this will be in the form of [intersection type]()

## Selectors

whereas the action puts data in the store; the selectors on the other hand act as getters of the slice of data that this reducer module handles

selectors are functions that will take at least one argument, the store, Then based on any other arguments like say an id, can be used to filter down the data retrieved.

## The Reducer

The reducer function, takes 2 arguments; the state and an action.
using the type property of the action, the reducer should then know how to modify the state.
such functionality could be implemented as a form of switch strcture such as a list of if statements or the more recommended approach which is to use the actual switch statement

### Adding the Reducer

- Define an interface for the data to be stored in the store
- Define an interface for the store state.
- write the actual reducer function, while applying the required type annotations as required.
