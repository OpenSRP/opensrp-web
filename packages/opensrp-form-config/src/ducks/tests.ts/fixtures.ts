export const fixManifestReleases = [
    {
        identifier: '1.0.11',
        json: '{"forms_version":"1.0.11","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:00:38.413+02:00',
        updatedAt: '2020-06-08T15:00:38.413+02:00',
    },
    {
        identifier: '1.0.12',
        json: '{"forms_version":"1.0.12","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:03:00.950+02:00',
        updatedAt: '2020-06-08T15:03:00.950+02:00',
    },
    {
        identifier: '1.0.16',
        json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.16',
        createdAt: '2020-06-09T14:40:04.189+02:00',
        updatedAt: '2020-06-09T14:40:04.189+02:00',
    },
];
