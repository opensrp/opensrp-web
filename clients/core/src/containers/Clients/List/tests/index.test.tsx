import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Provider } from 'react-redux';
import ConnectedClientsList, { ClientList } from '..';
import store from '../../../../store';
import reducer, { fetchClients, reducerName } from '../../../../store/ducks/clients';
import * as fixtures from '../../../../store/ducks/tests/fixtures';

reducerRegistry.register(reducerName, reducer);

jest.mock('../../../../configs/env');
describe('containers/clients/list/ClientList', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('renders without crashing', () => {
        const mock: jest.Mock = jest.fn();
        const opensrpServiceMock: jest.Mock = jest.fn();
        const props = {
            clientsArray: fixtures.clients,
            fetchClientsActionCreator: mock,
            location: mock,
            match: mock,
            opensrpService: opensrpServiceMock,
        };
        shallow(<ClientList {...props} />);
    });

    it('renders correctly', () => {
        const mock: jest.Mock = jest.fn();
        const opensrpServiceMock: jest.Mock = jest.fn();
        const props = {
            clientsArray: fixtures.clients,
            fetchClientsActionCreator: mock,
            location: mock,
            match: mock,
            opensrpService: opensrpServiceMock,
        };
        const wrapper = mount(<ClientList {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('renders correctly when clientsArray is an empty array', () => {
        const mock: jest.Mock = jest.fn();
        const opensrpServiceMock: jest.Mock = jest.fn();
        const props = {
            clientsArray: [],
            fetchClientsActionCreator: mock,
            location: mock,
            match: mock,
            opensrpService: opensrpServiceMock,
        };
        const wrapper = mount(<ClientList {...props} />);
        expect(toJson(wrapper.find('Ripple'))).toMatchSnapshot('Ripple Loader');
        wrapper.unmount();
    });

    it('works correctly with the redux store', () => {
        store.dispatch(fetchClients(fixtures.clients));
        const mock: jest.Mock = jest.fn();
        const opensrpServiceMock: jest.Mock = jest.fn();
        const props = {
            fetchClientsActionCreator: mock,
            location: mock,
            match: mock,
            opensrpService: opensrpServiceMock,
        };
        const wrapper = mount(
            <Provider store={store}>
                <ConnectedClientsList {...props} />
                );
            </Provider>,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('calls openSRPService with the correct params', () => {
        const mock: jest.Mock = jest.fn();
        const opensrpServiceMock: jest.Mock = jest.fn();
        const props = {
            fetchClientsActionCreator: mock,
            location: mock,
            match: mock,
            opensrpService: opensrpServiceMock,
        };
        const wrapper = mount(
            <Provider store={store}>
                <ConnectedClientsList {...props} />
                );
            </Provider>,
        );
        expect(opensrpServiceMock.mock.calls[0][0]).toEqual('https://test.smartregister.org/opensrp/rest/');
        expect(opensrpServiceMock.mock.calls[0][1]).toEqual('client/search');
        wrapper.unmount();
    });
});
