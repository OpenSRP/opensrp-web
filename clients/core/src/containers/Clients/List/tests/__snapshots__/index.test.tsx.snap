// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/clients/list/ClientList renders correctly 1`] = `
<ClientList
  clientsArray={
    Array [
      Object {
        "_id": "9b67a82d-dac7-40c0-85aa-e5976339a6b6",
        "_rev": "v1",
        "addresses": Array [
          Object {
            "addressFields": null,
            "addressType": "",
            "cityVillage": "Tha Luang",
            "country": null,
            "countyDistrict": null,
            "endDate": null,
            "geopoint": null,
            "latitude": null,
            "longitude": null,
            "postalCode": null,
            "preferred": null,
            "startDate": null,
            "stateProvince": null,
            "subDistrict": null,
            "subTown": null,
            "town": null,
          },
        ],
        "attributes": Object {
          "residence": "da765947-5e4d-49f7-9eb8-2d2d00681f65",
        },
        "baseEntityId": "71ad460c-bf76-414e-9be1-0d1b2cb1bce8",
        "birthdate": 7200000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557670951023,
        "deathdateApprox": false,
        "firstName": "Khumpai",
        "gender": "Male",
        "identifiers": Object {
          "opensrp_id": "11096120_family",
        },
        "lastName": "Family",
        "relationships": Object {
          "family_head": Array [
            "7d97182f-d623-4553-8651-5a29d2fe3f0b",
          ],
          "primary_caregiver": Array [
            "7d97182f-d623-4553-8651-5a29d2fe3f0b",
          ],
        },
        "serverVersion": 1557670950986,
        "type": "Client",
      },
      Object {
        "_id": "a30116d5-0612-419e-9b93-00c87df4ffbb",
        "_rev": "v1",
        "addresses": Array [],
        "attributes": Object {
          "residence": "da765947-5e4d-49f7-9eb8-2d2d00681f65",
        },
        "baseEntityId": "7d97182f-d623-4553-8651-5a29d2fe3f0b",
        "birthdate": 594777600000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557670951165,
        "deathdateApprox": false,
        "firstName": "Pramoj",
        "identifiers": Object {
          "opensrp_id": "11096120",
        },
        "lastName": "Khumpai",
        "middleName": "",
        "relationships": Object {
          "family": Array [
            "71ad460c-bf76-414e-9be1-0d1b2cb1bce8",
          ],
        },
        "serverVersion": 1557670950999,
        "type": "Client",
      },
      Object {
        "_id": "b23307c2-cbbb-481f-9cd7-f218564cc4bc",
        "_rev": "v1",
        "addresses": Array [
          Object {
            "addressFields": null,
            "addressType": "",
            "cityVillage": "Tha Luang",
            "country": null,
            "countyDistrict": null,
            "endDate": null,
            "geopoint": null,
            "latitude": null,
            "longitude": null,
            "postalCode": null,
            "preferred": null,
            "startDate": null,
            "stateProvince": null,
            "subDistrict": null,
            "subTown": null,
            "town": null,
          },
        ],
        "attributes": Object {
          "residence": "bd73f7d7-4387-4b6b-b632-acb03c4ea160",
        },
        "baseEntityId": "e540ac0e-d913-4898-846a-d3ee31863182",
        "birthdate": 7200000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557672169407,
        "deathdateApprox": false,
        "firstName": "Sepsook",
        "gender": "Male",
        "identifiers": Object {
          "opensrp_id": "11099629_family",
        },
        "lastName": "Family",
        "relationships": Object {
          "family_head": Array [
            "aaaca36a-a913-466b-9dd7-0c68db0f7c6f",
          ],
          "primary_caregiver": Array [
            "aaaca36a-a913-466b-9dd7-0c68db0f7c6f",
          ],
        },
        "serverVersion": 1557672169404,
        "type": "Client",
      },
      Object {
        "_id": "e5929901-9377-413c-850c-e41b48b34d5f",
        "_rev": "v1",
        "addresses": Array [],
        "attributes": Object {
          "residence": "bd73f7d7-4387-4b6b-b632-acb03c4ea160",
        },
        "baseEntityId": "aaaca36a-a913-466b-9dd7-0c68db0f7c6f",
        "birthdate": 868320000000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557672169419,
        "deathdateApprox": false,
        "firstName": "Udom",
        "identifiers": Object {
          "opensrp_id": "11099629",
        },
        "lastName": "Sepsook",
        "middleName": "",
        "relationships": Object {
          "family": Array [
            "e540ac0e-d913-4898-846a-d3ee31863182",
          ],
        },
        "serverVersion": 1557672169404,
        "type": "Client",
      },
      Object {
        "_id": "faa436bb-0aa5-4c5f-a2fe-be687e164205",
        "_rev": "v1",
        "addresses": Array [],
        "attributes": Object {
          "residence": "6f7ca772-b368-4c3d-bd9c-00aa698203ca",
        },
        "baseEntityId": "630739be-e0ba-4f28-a5e7-96d4960fe1a2",
        "birthdate": 1463097600000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557737485545,
        "deathdateApprox": false,
        "firstName": "Imalda",
        "gender": "Female",
        "identifiers": Object {
          "opensrp_id": "11117132",
        },
        "lastName": "DANE",
        "middleName": "",
        "relationships": Object {
          "family": Array [
            "97231120-8715-4131-b472-f71ecf230730",
          ],
        },
        "serverVersion": 1557737485527,
        "type": "Client",
      },
      Object {
        "_id": "0a5457f9-3e2e-4432-afae-75b57251651a",
        "_rev": "v1",
        "addresses": Array [
          Object {
            "addressFields": Object {
              "address2": "3",
              "street": "no street",
            },
            "addressType": "",
            "cityVillage": "where ever I am ",
            "country": null,
            "countyDistrict": null,
            "endDate": null,
            "geopoint": null,
            "latitude": null,
            "longitude": null,
            "postalCode": null,
            "preferred": null,
            "startDate": null,
            "stateProvince": null,
            "subDistrict": null,
            "subTown": null,
            "town": null,
          },
        ],
        "attributes": Object {
          "residence": "6f7ca772-b368-4c3d-bd9c-00aa698203ca",
        },
        "baseEntityId": "97231120-8715-4131-b472-f71ecf230730",
        "birthdate": 7200000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557737728445,
        "deathdateApprox": false,
        "firstName": "DANE",
        "gender": "Male",
        "identifiers": Object {
          "opensrp_id": "11117124_family",
        },
        "lastName": "Family",
        "relationships": Object {
          "family_head": Array [
            "945fc98d-4cce-412a-8327-ca2315efedf3",
          ],
          "primary_caregiver": Array [
            "945fc98d-4cce-412a-8327-ca2315efedf3",
          ],
        },
        "serverVersion": 1557737728441,
        "type": "Client",
      },
      Object {
        "_id": "6dab682a-0f31-4935-9114-c4d33d148617",
        "_rev": "v1",
        "addresses": Array [],
        "attributes": Object {
          "residence": "6f7ca772-b368-4c3d-bd9c-00aa698203ca",
        },
        "baseEntityId": "945fc98d-4cce-412a-8327-ca2315efedf3",
        "birthdate": 358560000000,
        "birthdateApprox": false,
        "clientApplicationVersion": 2,
        "clientDatabaseVersion": 2,
        "dateCreated": 1557737728487,
        "deathdateApprox": false,
        "firstName": "Peter",
        "identifiers": Object {
          "opensrp_id": "11117124",
        },
        "lastName": "DANE",
        "middleName": "",
        "relationships": Object {
          "family": Array [
            "97231120-8715-4131-b472-f71ecf230730",
          ],
        },
        "serverVersion": 1557737728441,
        "type": "Client",
      },
    ]
  }
  fetchClientsActionCreator={[MockFunction]}
  location={[MockFunction]}
  match={[MockFunction]}
  opensrpService={
    [MockFunction] {
      "calls": Array [
        Array [
          "client/search",
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  removeClients={[Function]}
  setTotalRecords={[Function]}
  totalRecords={0}
>
  <Ripple
    borderColor="#ff5d00"
    borderStyle="solid"
    borderWidth="4px"
    height="64px"
    minHeight="80vh"
    width="64px"
  >
    <div
      className="lds-ripple-parent"
      style={
        Object {
          "minHeight": "80vh",
        }
      }
    >
      <div
        className="lds-ripple"
        style={
          Object {
            "height": "64px",
            "width": "64px",
          }
        }
      >
        <div
          style={
            Object {
              "borderColor": "#ff5d00",
              "borderStyle": "solid",
              "borderWidth": "4px",
            }
          }
        />
        <div
          style={
            Object {
              "borderColor": "#ff5d00",
              "borderStyle": "solid",
              "borderWidth": "4px",
            }
          }
        />
      </div>
    </div>
  </Ripple>
</ClientList>
`;

exports[`containers/clients/list/ClientList renders correctly when clientsArray is an empty array: Ripple Loader 1`] = `
<Ripple
  borderColor="#ff5d00"
  borderStyle="solid"
  borderWidth="4px"
  height="64px"
  minHeight="80vh"
  width="64px"
>
  <div
    className="lds-ripple-parent"
    style={
      Object {
        "minHeight": "80vh",
      }
    }
  >
    <div
      className="lds-ripple"
      style={
        Object {
          "height": "64px",
          "width": "64px",
        }
      }
    >
      <div
        style={
          Object {
            "borderColor": "#ff5d00",
            "borderStyle": "solid",
            "borderWidth": "4px",
          }
        }
      />
      <div
        style={
          Object {
            "borderColor": "#ff5d00",
            "borderStyle": "solid",
            "borderWidth": "4px",
          }
        }
      />
    </div>
  </div>
</Ripple>
`;

exports[`containers/clients/list/ClientList works correctly with the redux store 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ClientList)
    fetchClientsActionCreator={[MockFunction]}
    location={[MockFunction]}
    match={[MockFunction]}
    opensrpService={
      [MockFunction] {
        "calls": Array [
          Array [
            "client/search",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <ClientList
      clientsArray={
        Array [
          Object {
            "_id": "9b67a82d-dac7-40c0-85aa-e5976339a6b6",
            "_rev": "v1",
            "addresses": Array [
              Object {
                "addressFields": null,
                "addressType": "",
                "cityVillage": "Tha Luang",
                "country": null,
                "countyDistrict": null,
                "endDate": null,
                "geopoint": null,
                "latitude": null,
                "longitude": null,
                "postalCode": null,
                "preferred": null,
                "startDate": null,
                "stateProvince": null,
                "subDistrict": null,
                "subTown": null,
                "town": null,
              },
            ],
            "attributes": Object {
              "residence": "da765947-5e4d-49f7-9eb8-2d2d00681f65",
            },
            "baseEntityId": "71ad460c-bf76-414e-9be1-0d1b2cb1bce8",
            "birthdate": 7200000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557670951023,
            "deathdateApprox": false,
            "firstName": "Khumpai",
            "gender": "Male",
            "identifiers": Object {
              "opensrp_id": "11096120_family",
            },
            "lastName": "Family",
            "relationships": Object {
              "family_head": Array [
                "7d97182f-d623-4553-8651-5a29d2fe3f0b",
              ],
              "primary_caregiver": Array [
                "7d97182f-d623-4553-8651-5a29d2fe3f0b",
              ],
            },
            "serverVersion": 1557670950986,
            "type": "Client",
          },
          Object {
            "_id": "a30116d5-0612-419e-9b93-00c87df4ffbb",
            "_rev": "v1",
            "addresses": Array [],
            "attributes": Object {
              "residence": "da765947-5e4d-49f7-9eb8-2d2d00681f65",
            },
            "baseEntityId": "7d97182f-d623-4553-8651-5a29d2fe3f0b",
            "birthdate": 594777600000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557670951165,
            "deathdateApprox": false,
            "firstName": "Pramoj",
            "identifiers": Object {
              "opensrp_id": "11096120",
            },
            "lastName": "Khumpai",
            "middleName": "",
            "relationships": Object {
              "family": Array [
                "71ad460c-bf76-414e-9be1-0d1b2cb1bce8",
              ],
            },
            "serverVersion": 1557670950999,
            "type": "Client",
          },
          Object {
            "_id": "b23307c2-cbbb-481f-9cd7-f218564cc4bc",
            "_rev": "v1",
            "addresses": Array [
              Object {
                "addressFields": null,
                "addressType": "",
                "cityVillage": "Tha Luang",
                "country": null,
                "countyDistrict": null,
                "endDate": null,
                "geopoint": null,
                "latitude": null,
                "longitude": null,
                "postalCode": null,
                "preferred": null,
                "startDate": null,
                "stateProvince": null,
                "subDistrict": null,
                "subTown": null,
                "town": null,
              },
            ],
            "attributes": Object {
              "residence": "bd73f7d7-4387-4b6b-b632-acb03c4ea160",
            },
            "baseEntityId": "e540ac0e-d913-4898-846a-d3ee31863182",
            "birthdate": 7200000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557672169407,
            "deathdateApprox": false,
            "firstName": "Sepsook",
            "gender": "Male",
            "identifiers": Object {
              "opensrp_id": "11099629_family",
            },
            "lastName": "Family",
            "relationships": Object {
              "family_head": Array [
                "aaaca36a-a913-466b-9dd7-0c68db0f7c6f",
              ],
              "primary_caregiver": Array [
                "aaaca36a-a913-466b-9dd7-0c68db0f7c6f",
              ],
            },
            "serverVersion": 1557672169404,
            "type": "Client",
          },
          Object {
            "_id": "e5929901-9377-413c-850c-e41b48b34d5f",
            "_rev": "v1",
            "addresses": Array [],
            "attributes": Object {
              "residence": "bd73f7d7-4387-4b6b-b632-acb03c4ea160",
            },
            "baseEntityId": "aaaca36a-a913-466b-9dd7-0c68db0f7c6f",
            "birthdate": 868320000000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557672169419,
            "deathdateApprox": false,
            "firstName": "Udom",
            "identifiers": Object {
              "opensrp_id": "11099629",
            },
            "lastName": "Sepsook",
            "middleName": "",
            "relationships": Object {
              "family": Array [
                "e540ac0e-d913-4898-846a-d3ee31863182",
              ],
            },
            "serverVersion": 1557672169404,
            "type": "Client",
          },
          Object {
            "_id": "faa436bb-0aa5-4c5f-a2fe-be687e164205",
            "_rev": "v1",
            "addresses": Array [],
            "attributes": Object {
              "residence": "6f7ca772-b368-4c3d-bd9c-00aa698203ca",
            },
            "baseEntityId": "630739be-e0ba-4f28-a5e7-96d4960fe1a2",
            "birthdate": 1463097600000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557737485545,
            "deathdateApprox": false,
            "firstName": "Imalda",
            "gender": "Female",
            "identifiers": Object {
              "opensrp_id": "11117132",
            },
            "lastName": "DANE",
            "middleName": "",
            "relationships": Object {
              "family": Array [
                "97231120-8715-4131-b472-f71ecf230730",
              ],
            },
            "serverVersion": 1557737485527,
            "type": "Client",
          },
          Object {
            "_id": "0a5457f9-3e2e-4432-afae-75b57251651a",
            "_rev": "v1",
            "addresses": Array [
              Object {
                "addressFields": Object {
                  "address2": "3",
                  "street": "no street",
                },
                "addressType": "",
                "cityVillage": "where ever I am ",
                "country": null,
                "countyDistrict": null,
                "endDate": null,
                "geopoint": null,
                "latitude": null,
                "longitude": null,
                "postalCode": null,
                "preferred": null,
                "startDate": null,
                "stateProvince": null,
                "subDistrict": null,
                "subTown": null,
                "town": null,
              },
            ],
            "attributes": Object {
              "residence": "6f7ca772-b368-4c3d-bd9c-00aa698203ca",
            },
            "baseEntityId": "97231120-8715-4131-b472-f71ecf230730",
            "birthdate": 7200000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557737728445,
            "deathdateApprox": false,
            "firstName": "DANE",
            "gender": "Male",
            "identifiers": Object {
              "opensrp_id": "11117124_family",
            },
            "lastName": "Family",
            "relationships": Object {
              "family_head": Array [
                "945fc98d-4cce-412a-8327-ca2315efedf3",
              ],
              "primary_caregiver": Array [
                "945fc98d-4cce-412a-8327-ca2315efedf3",
              ],
            },
            "serverVersion": 1557737728441,
            "type": "Client",
          },
          Object {
            "_id": "6dab682a-0f31-4935-9114-c4d33d148617",
            "_rev": "v1",
            "addresses": Array [],
            "attributes": Object {
              "residence": "6f7ca772-b368-4c3d-bd9c-00aa698203ca",
            },
            "baseEntityId": "945fc98d-4cce-412a-8327-ca2315efedf3",
            "birthdate": 358560000000,
            "birthdateApprox": false,
            "clientApplicationVersion": 2,
            "clientDatabaseVersion": 2,
            "dateCreated": 1557737728487,
            "deathdateApprox": false,
            "firstName": "Peter",
            "identifiers": Object {
              "opensrp_id": "11117124",
            },
            "lastName": "DANE",
            "middleName": "",
            "relationships": Object {
              "family": Array [
                "97231120-8715-4131-b472-f71ecf230730",
              ],
            },
            "serverVersion": 1557737728441,
            "type": "Client",
          },
        ]
      }
      fetchClientsActionCreator={[Function]}
      getTotalRecords={[Function]}
      location={[MockFunction]}
      match={[MockFunction]}
      opensrpService={
        [MockFunction] {
          "calls": Array [
            Array [
              "https://test.smartregister.org/opensrp/rest/",
              "client/search",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      removeClients={[Function]}
      setTotalRecords={[Function]}
      totalRecords={0}
    >
      <Ripple
        borderColor="#ff5d00"
        borderStyle="solid"
        borderWidth="4px"
        height="64px"
        minHeight="80vh"
        width="64px"
      >
        <div
          className="lds-ripple-parent"
          style={
            Object {
              "minHeight": "80vh",
            }
          }
        >
          <div
            className="lds-ripple"
            style={
              Object {
                "height": "64px",
                "width": "64px",
              }
            }
          >
            <div
              style={
                Object {
                  "borderColor": "#ff5d00",
                  "borderStyle": "solid",
                  "borderWidth": "4px",
                }
              }
            />
            <div
              style={
                Object {
                  "borderColor": "#ff5d00",
                  "borderStyle": "solid",
                  "borderWidth": "4px",
                }
              }
            />
          </div>
        </div>
      </Ripple>
    </ClientList>
  </Connect(ClientList)>
  );
</Provider>
`;
